Author: Uwe Augustin <Info@FlapsOnline.de>
Date:   Thu Jan 06 10:00:00 2025 +0200

    use CriticalSection on windows - not std::mutex
    #include "../../../../src/thread/CriticalSection.hxx"

diff --git a/src/mutex.cpp b/src/mutex.cpp
--- a/src/mutex.cpp
+++ b/src/mutex.cpp
@@ -29,9 +29,16 @@ #include <mutex>
 
 #include "proj.h"
 #include "proj_internal.h"
 
-static std::recursive_mutex core_lock;
+#if defined(_WIN32) && !defined(_MSC_VER)
+  #include "../../../../src/thread/CriticalSection.hxx"
+  using Mutex = CriticalSection;
+  using RecursiveMutex = CriticalSection;
+  static RecursiveMutex core_lock;
+#else
+  static std::recursive_mutex core_lock;
+#endif
 
 /************************************************************************/
 /*                          pj_acquire_lock()                           */
 /*                                                                      */
diff --git a/src/networkfilemanager.cpp b/src/networkfilemanager.cpp
--- a/src/networkfilemanager.cpp
+++ b/src/networkfilemanager.cpp
@@ -76,7 +76,16 @@ #include <time.h>
 
 //! @cond Doxygen_Suppress
 
+#if defined(_WIN32) && !defined(_MSC_VER)
+# include "../../../../src/thread/CriticalSection.hxx"
+  using Mutex = CriticalSection;
+  using RecursiveMutex = CriticalSection;
+#else
+  using Mutex = std::mutex;
+  using RecursiveMutex = std::recursive_mutex;
+#endif
+
 #define STR_HELPER(x) #x
 #define STR(x) STR_HELPER(x)
 
 using namespace NS_PROJ::internal;
@@ -143,6 +152,6 @@      };
 
     lru11::Cache<
-        Key, std::shared_ptr<std::vector<unsigned char>>, std::mutex,
+        Key, std::shared_ptr<std::vector<unsigned char>>, Mutex,
         std::unordered_map<
             Key,
             typename std::list<lru11::KeyValuePair<
@@ -167,6 +176,6 @@      void clearMemoryCache();
 
   private:
-    lru11::Cache<std::string, FileProperties, std::mutex> cache_{};
+    lru11::Cache<std::string, FileProperties, Mutex> cache_{};
 };
 
 // ---------------------------------------------------------------------------
diff --git a/src/transformations/gridshift.cpp b/src/transformations/gridshift.cpp
index 260b5b05..042a09b5 100644
--- a/src/transformations/gridshift.cpp
+++ b/src/transformations/gridshift.cpp
@@ -46,8 +46,13 @@
 #include <utility>
 
 PROJ_HEAD(gridshift, "Generic grid shift");
-
-static std::mutex gMutex{};
+#if defined(_WIN32) && !defined(_MSC_VER)
+# include "../../../../src/thread/CriticalSection.hxx"
+using Mutex = CriticalSection;
+#else
+using Mutex = std::mutex;
+#endif
+static Mutex gMutex{};
 // Map of (name, isProjected)
 static std::map<std::string, bool> gKnownGrids{};
 
@@ -1008,6 +1013,6 @@ PJ *PJ_TRANSFORMATION(gridshift, 0) {
 // ---------------------------------------------------------------------------
 
 void pj_clear_gridshift_knowngrids_cache() {
-    std::lock_guard<std::mutex> lock(gMutex);
+    std::lock_guard<Mutex> lock(gMutex);
     gKnownGrids.clear();
 }
diff --git a/src/transformations/hgridshift.cpp b/src/transformations/hgridshift.cpp
index 4be74159..4355899a 100644
--- a/src/transformations/hgridshift.cpp
+++ b/src/transformations/hgridshift.cpp
@@ -11,7 +11,13 @@
 
 PROJ_HEAD(hgridshift, "Horizontal grid shift");
 
-static std::mutex gMutexHGridShift{};
+#if defined(_WIN32) && !defined(_MSC_VER)
+# include "../../../../src/thread/CriticalSection.hxx"
+using Mutex = CriticalSection;
+#else
+using Mutex = std::mutex;
+#endif
+static Mutex gMutexHGridShift{};
 static std::set<std::string> gKnownGridsHGridShift{};
 
 using namespace NS_PROJ;
@@ -206,6 +212,6 @@ PJ *PJ_TRANSFORMATION(hgridshift, 0) {
 }
 
 void pj_clear_hgridshift_knowngrids_cache() {
-    std::lock_guard<std::mutex> lock(gMutexHGridShift);
+    std::lock_guard<Mutex> lock(gMutexHGridShift);
     gKnownGridsHGridShift.clear();
 }
diff --git a/src/transformations/vgridshift.cpp b/src/transformations/vgridshift.cpp
index 6460bfd6..de15fabe 100644
--- a/src/transformations/vgridshift.cpp
+++ b/src/transformations/vgridshift.cpp
@@ -11,7 +11,13 @@
 
 PROJ_HEAD(vgridshift, "Vertical grid shift");
 
-static std::mutex gMutexVGridShift{};
+#if defined(_WIN32) && !defined(_MSC_VER)
+# include "../../../../src/thread/CriticalSection.hxx"
+using Mutex = CriticalSection;
+#else
+using Mutex = std::mutex;
+#endif
+static Mutex gMutexVGridShift{};
 static std::set<std::string> gKnownGridsVGridShift{};
 
 using namespace NS_PROJ;
@@ -246,6 +252,6 @@ PJ *PJ_TRANSFORMATION(vgridshift, 0) {
 }
 
 void pj_clear_vgridshift_knowngrids_cache() {
-    std::lock_guard<std::mutex> lock(gMutexVGridShift);
+    std::lock_guard<Mutex> lock(gMutexVGridShift);
     gKnownGridsVGridShift.clear();
 }
diff --git a/src/iso19111/factory.cpp b/src/iso19111/factory.cpp
--- a/src/iso19111/factory.cpp
+++ b/src/iso19111/factory.cpp
@@ -94,6 +94,12 @@
 #define REOPEN_SQLITE_DB_AFTER_FORK
 #endif
 
+#if defined(_WIN32) && !defined(_MSC_VER)
+# include "../../../../src/thread/CriticalSection.hxx"
+  using Mutex = CriticalSection;
+#else
+  using Mutex = std::mutex;
+#endif
 using namespace NS_PROJ::internal;
 using namespace NS_PROJ::common;
 
@@ -585,7 +591,7 @@ class SQLiteHandleCache {
     bool firstTime_ = true;
 #endif
 
-    std::mutex sMutex_{};
+    Mutex sMutex_{};
 
     // Map dbname to SQLiteHandle
     lru11::Cache<std::string, std::shared_ptr<SQLiteHandle>> cache_{};
@@ -614,7 +620,7 @@ SQLiteHandleCache &SQLiteHandleCache::get() {
 // ---------------------------------------------------------------------------
 
 void SQLiteHandleCache::clear() {
-    std::lock_guard<std::mutex> lock(sMutex_);
+    std::lock_guard<Mutex> lock(sMutex_);
     cache_.clear();
 }
 
@@ -622,7 +628,7 @@ void SQLiteHandleCache::clear() {
 
 std::shared_ptr<SQLiteHandle>
 SQLiteHandleCache::getHandle(const std::string &path, PJ_CONTEXT *ctx) {
-    std::lock_guard<std::mutex> lock(sMutex_);
+    std::lock_guard<Mutex> lock(sMutex_);
 
 #ifdef REOPEN_SQLITE_DB_AFTER_FORK
     if (firstTime_) {
@@ -645,7 +651,7 @@ SQLiteHandleCache::getHandle(const std::string &path, PJ_CONTEXT *ctx) {
 // ---------------------------------------------------------------------------
 
 void SQLiteHandleCache::invalidateHandles() {
-    std::lock_guard<std::mutex> lock(sMutex_);
+    std::lock_guard<Mutex> lock(sMutex_);
     const auto lambda =
         [](const lru11::KeyValuePair<std::string, std::shared_ptr<SQLiteHandle>>
                &kvp) { kvp.value->invalidate(); };
